manager/orchestrator/globaljob/reconciler.go

```diff
@@ -0,0 +1,172 @@
+package globaljob
+
+import (
+	"fmt"
+
+	"github.com/docker/swarmkit/api"
+	"github.com/docker/swarmkit/manager/constraint"
+	"github.com/docker/swarmkit/manager/orchestrator"
+	"github.com/docker/swarmkit/manager/state/store"
+	gogotypes "github.com/gogo/protobuf/types"
+)
+
+type reconciler interface {
+	ReconcileService(id string) error
+}
+
+type reconcilerObj struct {
+	store *store.MemoryStore
+}
+
+func newReconciler(store *store.MemoryStore) reconciler {
+	return &reconcilerObj{
+		store: store,
+	}
+}
+
+func (r *reconcilerObj) ReconcileService(id string) error {
+	var (
+		service *api.Service
+		cluster *api.Cluster
+		tasks   []*api.Task
+		nodes   []*api.Node
+	)
+
+	// we need to first get the latest iteration of the service, its tasks, and
+	// the nodes in the cluster.
+	r.store.View(func(tx store.ReadTx) {
+		service = store.GetService(tx, id)
+		if service == nil {
+			return
+		}
+
+		// getting tasks with FindTasks should only return an error if we've
+		// made a mistake coding; there's no user-input or even reasonable
+		// system state that can cause it. If it returns an error, we'll just
+		// panic and crash.
+		var err error
+		tasks, err = store.FindTasks(tx, store.ByServiceID(id))
+		if err != nil {
+			panic(fmt.Sprintf("error getting tasks: %v", err))
+		}
+
+		// same as with FindTasks
+		nodes, err = store.FindNodes(tx, store.All)
+		if err != nil {
+			panic(fmt.Sprintf("error getting nodes: %v", err))
+		}
+
+		clusters, _ := store.FindClusters(tx, store.All)
+		if len(clusters) == 1 {
+			cluster = clusters[0]
+		} else if len(clusters) > 1 {
+			panic("there should never be more than one cluster object")
+		}
```

Comment:
Duplicated from https://github.com/docker/swarmkit/pull/2883/files#diff-ddae210b97adba6178634ae53b773f2a. Maybe that could be put in some helper func in the `store` package, say  `store.FindCluter(store.ReadTx) Cluster`?